assets/src/components/template-inserter/index.js

```diff
@@ -0,0 +1,163 @@
+/**
+ * WordPress dependencies
+ */
+import { __, sprintf } from '@wordpress/i18n';
+import { Dropdown, IconButton, Spinner } from '@wordpress/components';
+import { Component } from '@wordpress/element';
+import { dispatch } from '@wordpress/data';
+import { parse, createBlock } from '@wordpress/blocks';
+const { apiFetch } = wp;
+
+const blocksRestBase = 'blocks';
+
+/**
+ * Internal dependencies
+ */
+import BlockPreview from './block-preview';
+import pageIcon from './icon';
+
+class TemplateInserter extends Component {
+	constructor() {
+		super( ...arguments );
+
+		this.onToggle = this.onToggle.bind( this );
+
+		this.state = {
+			reusableBlocks: null,
+		};
+	}
+
+	componentDidMount() {
+		// This is used for making sure that once the response finishes the component actually still exists.
+		this.isComponentMounted = true;
+		this.getReusableBlocks();
+	}
+
+	componentWillUnmount() {
+		this.isComponentMounted = false;
+	}
+
+	onToggle( isOpen ) {
+		const { onToggle } = this.props;
+
+		// Surface toggle callback to parent component
+		if ( onToggle ) {
+			onToggle( isOpen );
+		}
+	}
+
+	getReusableBlocks() {
+		if ( ! this.isComponentMounted ) {
+			return;
+		}
+
+		if ( null !== this.state.reusableBlocks ) {
+			this.setState( { reusableBlocks: null } );
+		}
+
+		// @todo We only need reusable blocks that can be used by AMP Stories.
+		const blocksRequest = this.lastRequest = apiFetch( {
```

Comment:
_Thinking out loud here_

One downside of doing the API request in the component is that we can't easily use the component twice without duplicating requests.

I was looking a bit at the `Inserter` and `InserterMenu` components in Gutenberg and how they use [`fetchReusableBlocks`](https://github.com/WordPress/gutenberg/blob/828fe40798b4d9b4a4d344b68eb0e2f62632ee8c/packages/editor/src/store/effects/reusable-blocks.js#L41-L99) and some other functions to do the data fetching. Are there perhaps some commonalities that we could re-use (maybe in combination with an [`api-fetch` middleware](https://wordpress.org/gutenberg/handbook/designers-developers/developers/packages/packages-api-fetch/#middlewares) to do the filtering) or copy over to our own data store?

> We only need reusable blocks that can be used by AMP Stories.

This goes the other way around too: In all other editors we don't want to show AMP Stories templates. `?search` doesn't work for that. An easy solution could be using a taxonomy behind the scenes. Post meta could work too, but doesn't scale well.

Reply:
> ?search doesn't work for that. An easy solution could be using a taxonomy behind the scenes. Post meta could work too, but doesn't scale well.

Agreed, `?search` was added more of a placeholder for preventing fetching all the reusable blocks which would then just cause errors on insertion or due to unregistered block types. I was thinking that implementing taxonomy / meta could be done either in #2010 / #2011 since those PR-s involve creating template posts.

>Are there perhaps some commonalities that we could re-use (maybe in combination with an api-fetch middleware to do the filtering) or copy over to our own data store?

Actually, we could also use the default experimental dispatch/selector for getting all reusable blocks and filter those. I thought initially that this might break the logic due to potentially including unregistered blocks, however, it looks like allowing `core/template` as an allowed block should resolve that issue. I'll make the change to test it out and then if it seems like it would make sense to manage the reusable blocks in our own data store then we can rework that later. Thoughts?

Reply:
This would mean though that we'd need to rework it anyway once taxonomy comes in. So yes, perhaps it does make sense to move it to our own data store.

Reply:
Alternatively we could just filter by having `amp-story-page` as the main block of the template, and not allow anything else, maybe just that could make sense for now, too. Thoughts?

Reply:
Makes sense to me. Let's stick with the current approach and iterate in follow-up PRs.